src/wallet/spend.cpp

```diff
@@ -799,6 +815,83 @@ util::Result<SelectionResult> SelectCoins(const CWallet& wallet, CoinsResult& av
     return op_selection_result;
 }
 
+/** Returns UTXO targets and compute their current count from available coins */
+std::vector<UtxoTarget> UtxoTargetsFromJson(const UniValue& utxo_targets_json, const CoinsResult& available_coins)
+{
+    std::vector<UtxoTarget> utxo_targets;
+    for (const auto& bucket : utxo_targets_json.getValues()) {
+        UtxoTarget target = {
+            bucket["start_satoshis"].getInt<CAmount>(),
+            bucket["end_satoshis"].getInt<CAmount>(),
+            bucket["target_utxo_count"].getInt<uint32_t>()};
+        utxo_targets.push_back(target);
+    }
+    for (const auto& [type, outputs] : available_coins.coins) {
+        for (const wallet::COutput& output : outputs) {
+            for (UtxoTarget& target : utxo_targets) {
+                if (output.txout.nValue >= target.start_satoshis && output.txout.nValue <= target.end_satoshis) {
+                    target.current_utxo_count++;
+                    break;
+                }
+            }
+        }
+    }
+
+    return utxo_targets;
+}
+
+/** Returns a random change amount in the range of the most depleted Utxo bucket and sets `capacity`
+ * to the capacity of that change target, if any.
+ */
+std::optional<CAmount> GenerateChangeTargetFromUtxoTargets(const std::vector<UtxoTarget>& utxo_targets, const CAmount change_fee, double& capacity, FastRandomContext& rng)
```

Comment:
I'm not sure this is really what we want. I think this function should try to refill multiple buckets at once (when amounts allow it), not just the most depleted one:

- Check which utxo buckets should be refilled:
  - Initialize a `to_refill` list of `(bucket, target_quantity)`
  - For each utxo bucket:
    - If the feerate is low (below a to-be-defined threshold) and the bucket is less than 70% full (or a to-be-defined threshold that could be configurable):
      - Add this bucket to `to_refill` with a `target_quantity` of `target_utxo_count - current_utxo_count` (when feerate is low, we'd like to refill as much as possible)
    - If the feerate is higher than our threshold and the bucket is less than 30% full (or a to-be-defined threshold that could be configurable), we refill less aggressively:
      - Add this bucket to `to_refill` with a `target_quantity` that we scale (somehow) based on feerate and how many utxos we're missing
  - Order `to_refill` by `target_quantity`, descending (we'll want to refill the buckets that are missing the most utxos first)
- Check the utxos that we have outside of our buckets: we ideally want to spend those utxos to refill our buckets. We have no guarantee that those are the ones that will be used by the coin selection, but we can use their total amount as the maximum amount of funds we allocate to refill our buckets. This part is a bit fuzzy for me right now, I'm not sure this is the best approach, it definitely deserves more thoughts.
  - Iterate over those utxos and add their amount to obtain `utxos_outside_buckets_total_amount`
- Compute the change target based on `to_refill` and `utxos_outside_buckets_total_amount`:
  - Initialize `change_target` to `0`
  - Iterate over `to_refill`, and for each bucket that needs to be refilled:
    - Decide how much of the `target_quantity` we want to refill based on `utxos_outside_buckets_total_amount` and the current `change_target`
    - Add the corresponding amount to `change_target`
    - That step probably has a lot of issues as well and deserves more thought. If we have a lot of utxos outside of our buckets, we may end up targeting a very large change amount and thus create a huge transaction, which is generally undesirable, but maybe desirable when our buckets are close to being empty. Maybe it makes sense to bound the `change_target` to a (small) multiplier of the funding amount? Or have a very different behavior when the feerate is very low, because when that happens we may want to fully refill our buckets?

As you can see, this is still very early discussion on what the algorithm should look like. It feels like we're still trying to understand the pitfalls we want to avoid we should try to write the algorithm in pseudo-code first in order to converge on an initial version (maybe to be detailed and discussed in a delving bitcoin post).

Also, when we don't want to actively refill buckets, we should target a changeless transaction, I'm not sure that is done right now?

Reply:
I think it is unavoidable that refilling buckets cannibalizes utxos from other buckets unless there is a pool of utxos larger than the largest target bucket to pick from. The goal though is that in the long run the system wastes as little value from fees and over payment as possible in aggregate.

This is how I think the process as implemented currently ([85a049d](https://github.com/bitcoin/bitcoin/pull/29442/commits/85a049ddceb854106efea8316462e356d9b88b21)) should work: 

Because we always set our minimum change value to be *at least* as large as one of our target buckets that is depleted,  all of the wallet utxos are either in the range of one of our target buckets or larger than one of our target buckets.

A changeless BnB solution is found when:
1) there is a single utxo in the corresponding target bucket that does not overpay by too much in fees
2) a combination of multiple utxos smaller than the target bucket exist that do not overpay by too much in fees

A CoinGrinder solution is preferred when the cost of fees and over payment of the best BnB solution is more than the fees from using:
1) a single larger utxo with change outputs 
2) a combination of multiple smaller utxos and change outputs

The largest wallet utxo will hold the residual value after initially refilling the buckets from cold storage. This residual funding utxo will be used as a single input when:
1) this utxo is available and CoinGrinder selects it to refill buckets opportunistically 
2) fees are low, or buckets are severely depleted, we force this input to be selected to refill buckets. 

When a large residual value utxo is used to refill all buckets via change outputs, another large residual utxo may also be created. That residual value utxo can not be used again until the tx that uses it confirms. This seems like an area that can be optimized. Perhaps a separate automatic refill transaction that confirms quickly makes more sense than using it for opportunistic bucket refills.

Eventually the largest utxo in the wallet will be in the range of the target buckets. The utxo set will then no longer be able to refill buckets without also depleting other buckets. This should be a signal to refill the wallet from cold storage.

Reply:
> Also, when we don't want to actively refill buckets, we should target a changeless transaction, I'm not sure that is done right now?

The current system should always prefer changeless transactions unless, for example, an exact match with more inputs and no change is more wasteful than selecting fewer inputs that generate a change output.